src/js/protocols/stm32.js

```diff
@@ -85,49 +88,120 @@ STM32_protocol.prototype.connect = function (port, baud, hex, options, callback)
                         }
                     });
                 });
+            });
+        };
+
+        var onConnectHandler = function () {
+
+            GUI.log(i18n.getMessage('apiVersionReceived', [CONFIG.apiVersion]));
+
+            if (semver.lt(CONFIG.apiVersion, "1.42.0")) {
+
+                self.msp_connector.disconnect(function (disconnectionResult) {
+
+                    // need some time for the port to be closed, serial port does not open if tried immediately
+                    setTimeout(legacyRebootAndFlash, 500);
+                });
             } else {
-                GUI.log(i18n.getMessage('serialPortOpenFail'));
+                console.log('Looking for capabilities via MSP');
+
+                MSP.send_message(MSPCodes.MSP_BOARD_INFO, false, false, function () {
+                    var rebootMode = 0; // FIRMWARE
+                    if (bit_check(CONFIG.targetCapabilities, FC.TARGET_CAPABILITIES_FLAGS.HAS_FLASH_BOOTLOADER)) {
+                        // Board has flash bootloader
+                        GUI.log(i18n.getMessage('deviceRebooting_flashBootloader'));
+                        console.log('flash bootloader detected');
+                        rebootMode = 4; // MSP_REBOOT_BOOTLOADER_FLASH
+                    } else {
+                        GUI.log(i18n.getMessage('deviceRebooting_romBootloader'));
+                        console.log('no flash bootloader detected');
+                        rebootMode = 1; // MSP_REBOOT_BOOTLOADER_ROM;
+                    }
+
+                    var buffer = [];
+                    buffer.push8(rebootMode);
+                    MSP.send_message(MSPCodes.MSP_SET_REBOOT, buffer, function() {
+
+                        // if firmware doesn't flush MSP/serial send buffers and gracefully shutdown VCP connections we won't get a reply, so don't wait for it.
+
+                        self.msp_connector.disconnect(function (disconnectionResult) {
+                            if (disconnectionResult) {
+                                // delay to allow board to boot in bootloader mode
+                                // required to detect if a DFU device appears
+                                setTimeout(startFlashing, 1000);
+                            } else {
+                                GUI.connect_lock = false;
+                            }
+                        });
+
+                    }, function () {
+                        console.log('Reboot request recevied by device');
+                    });
+                });
             }
-        });
+        }
+
+        var onTimeoutHandler = function() {
+            GUI.connect_lock = false;
+            console.log('Looking for capabilities via MSP failed');
+
+            TABS.firmware_flasher.flashingMessage(i18n.getMessage('stm32RebootingToBootloaderFailed'), TABS.firmware_flasher.FLASH_MESSAGE_TYPES.INVALID);
+        }
+
+        var onFailureHandler = function() {
```

Comment:
![Codacy](https://app.codacy.com/assets/images/favicon.png) Issue found: [Combine this with the previous 'var' statement.](https://app.codacy.com/gh/emuflight/EmuConfigurator/pullRequest?prid=5626707)